source/rust_verify_test/tests/loops.rs

```diff
@@ -983,3 +983,231 @@ test_verify_one_file! {
+                invariant
+                    n == iter.cur * 3,
+                    end == 10,
+            {
+                assert(x < 10);
+                assert(x == iter.cur);
+                n += 3;
+            }
+            assert(n == 30);
+        }
+
+        fn test_loop_fail() {
+            let mut n: u64 = 0;
+            let mut end = 10;
+            for x in iter: 0..end
+                invariant
+                    n == iter.cur * 3,
+                    end == 10,
+            {
+                assert(x < 10); // FAILS
+                assert(x == iter.cur);
+                n += 3;
+                end = end + 0; // causes end to be non-constant, so loop needs more invariants
+            }
+        }
+    } => Err(e) => assert_one_fails(e)
+}
+
+test_verify_one_file! {
+    #[test] for_loop3 verus_code! {
+        use vstd::prelude::*;
+        fn test_loop(n: u32) -> (v: Vec<u32>)
+            ensures
+                v.len() == n,
+                forall|i: int| 0 <= i < n ==> v[i] == i,
+        {
+            let mut v: Vec<u32> = Vec::new();
+            for i in iter: 0..n
+                invariant
+                    v@ =~= iter@,
+            {
+                v.push(i);
+            }
+            v
+        }
+
+        fn test_loop_fail(n: u32) -> (v: Vec<u32>)
+            ensures
+                v.len() == n,
+                forall|i: int| 0 <= i < n ==> v[i] == i,
+        {
+            let mut v: Vec<u32> = Vec::new();
+            for i in iter: 0..n
+                invariant
+                    v@ =~= iter@, // FAILS
+            {
+                v.push(0);
+            }
+            v
+        }
+    } => Err(e) => assert_one_fails(e)
+}
+
+test_verify_one_file! {
+    #[test] for_loop_vec_custom_iterator verus_code! {
+        use vstd::prelude::*;
+        pub struct VecIterCopy<'a, T: 'a> {
+            pub vec: &'a Vec<T>,
+            pub cur: usize,
+        }
+
+        impl<'a, T: Copy> Iterator for VecIterCopy<'a, T> {
+            type Item = T;
+            fn next(&mut self) -> (item: Option<T>)
+                ensures
+                    self.vec == old(self).vec,
+                    old(self).cur < self.vec.len() ==> self.cur == old(self).cur + 1,
+                    old(self).cur < self.vec.len() ==> item == Some(self.vec[old(self).cur as int]),
+                    old(self).cur >= self.vec.len() ==> item.is_none() && self.cur == old(self).cur,
+            {
+                if self.cur < self.vec.len() {
+                    let item = self.vec[self.cur];
+                    self.cur = self.cur + 1;
+                    Some(item)
+                } else {
+                    None
+                }
+            }
+        }
+
+        pub struct VecGhostIterCopy<'a, T> {
+            pub seq: Seq<T>,
+            pub cur: int,
+            pub unused_phantom_dummy: &'a Vec<T>, // use the 'a parameter, which I don't even want; ugh
```

Comment:
What's this comment about?

Reply:
I fixed this to use an actual `PhantomData` value